rfcs/ion_templates.md

```diff
@@ -0,0 +1,472 @@
+
+Unlike `symbol`s, `template`s are a system-level encoding detail, not a user-level type.
+
+While `template`s offer a superset of `symbol`s’ functionality and could replace them wholesale, this document proposes adding them alongside `symbol`s to preserve backwards compatibility.
+
+
+## New system symbols
+
+Support for templates requires three new symbols.
+
+Two for [defining a template](#defining-a-template):
+* `$ion_template_table`
+* `templates`
+
+And one to support the templates' [text encoding within S-Expressions](#text-encoding-within-s-expressions):
+* `$ion_template`
+
+
+
+## Defining a template
+
+Template tables are defined analogously to symbol tables, by adding an `$ion_template_table`-annotated struct to the value stream.
+
+```js
+$ion_template_table::{
+  templates : [
+    "sassafras", // template 0
+    1970-01-01T00:00:00Z, // template 1
+    3.14159265358979323846264338323 // template 2
+  ]
+}
+```
+
+Composite values (i.e. containers) can leave nested values unspecified by using the sentinel value, `@_`.
+
+```js
+$ion_template_table::{
+  templates : [{ // template 0
+    make: @_,
+    model: @_,
+    year: @_,
+    frame: sedan,
+    numberOfWheels: 4,
+    transmission: automatic,
+    hasAirbags: true
+  }]
+}
+```
+
+Template definitions can refer to other templates, but only those templates that have already been defined (i.e. templates with a lower ID) in order to cheaply avoid cycles.
+
+
+## Invoking a template
+
+### Invocation syntax
+
+```js
+// Invoke template 0 with the following parameters
+@0("Toyota", "Camry", 2017)
+```
+
+Parameters to a template invocation can themselves be template invocations. There is no restriction on which templates can be invoked as parameters.
+
+### Expanded data
+
+```js
+{ 
+  make: "Toyota",
+  model: "Camry",
+  year: 2017,
+  frame: "sedan",
+  numberOfWheels: 4,
+  transmission: "automatic",
+  hasAirbags: true
+}
+```
+
+In binary Ion, the template invocation requires 18 bytes to encode while the expanded data would require 42 bytes.
+
+## Binary layout
+
+See the [Ion 1.0 binary encoding spec](http://amzn.github.io/ion-docs/docs/binary.html#typed-value-formats) for a detailed explanation of the terms used in the diagram below.
+
+```
+            7       4 3       0
+           +---------+---------+
+Template   |   15    |    L    |
+           +---------+---------+======+
+           :     length [VarUInt]     :
+           +--------------------------+
+           |     template ID [VarUInt]|
+           +--------------------------+
+           :           value          :
+           +==========================+
+                         ⋮
+```
+
+* `L` values of `0` and `null` are available to be used as sentinels. 
+* This document treats `L`=0 as a template blank: `@_`
+* `L`=14 is a variable `length` encoding.
+* `L`=15 is reserved for future specs to accommodate not-yet-designed template invocation methods. (e.g. “splicing”)
```

Comment:
The problem I have with this definition is that it implies that `L=15` is reserved for just template related things, and if it is, then my largest problem with this proposal is that it "consumes" too much of the encoding space for future expansion.